.github/workflows/benchmark.yml

```diff
@@ -1,8 +1,7 @@
 name: Benchmark
 on:
-  pull_request:
-    branches: [master]
-    types: [opened, reopened, synchronize]
+  push:
+    branches: [master, staging, trying]
```

Comment:
This wont work because the base for the benchmark is `master` itself.

Reply:
So benchmark currently takes ~30 minutes, 14 minutes for master and 14 minutes for PR run.  Even the 14 minute run seems like a pretty long time to go, compared to the value we are getting.

I think we should do some combination of
- Remove the benchmark pipeline for now
- Re-evaluate our benchmarks if we can make them faster, generally
- Potentially identify a subset of benchmarks that are important for CI
- Identify alternative ways of tracking performance than running both master and PR

Reply:
Moving this comment here to keep the conversation in one place

@pksunkara 
> Why are we deleting this? I consider benchmarking every commit as important as coverage report and linting every commit. We just don't have good third party tools to properly consume this information yet. But I plan to gather the data and use it.
> 
> If need be, we can move this to master branch and compare by trying to checkout the previous commit on master. But if we are doing coverage on every PR, we should follow the same here. It's not like a PR is blocked on this or coverage.

Reply:
> Why are we deleting this? I consider benchmarking every commit as important as coverage report and linting every commit. We just don't have good third party tools to properly consume this information yet. But I plan to gather the data and use it.

If we aren't using the information yet, then its going to waste.  Let's reconsider this if/when we do use this information.

Even if we can, we need to consider the role of the information and how balance cost and benefit.  The cost is taking up an executor for 30min any time it runs (for `master`, its every update for every PR).  Most PRs will be performance neutral.  We can get most of the value by tracking `master`'s performance to see which commit introduced a slow down.

For me, I'd much rather we track a PRs impact on code bloat.  That is a metric a lot of users are complaining about and it is an easy one to negatively impact.

> . It's not like a PR is blocked on this or coverage.

It takes up one of the runners towards our cap, so they slow down progress, even if they don't gate changes.

Reply:
Very much agree with this sentiment. The benefit (catching performance regressions) must outweigh the costs (CI time / runner allocation).

I think once we have tooling to utilize benchmarks, and alert or fail CI on statistically significant regressions (which in and of itself might be a hard thing to do...do GH action runners suffer from noisy neighbors, etc.?) then we can (re-)enable them for a specific subset of PRs (release, etc.).

I say this [having caught big-ish performance regressions from small commits before](https://github.com/clap-rs/clap/pull/1436#issuecomment-480003706), so knowing that I very much care catching those. But even in the linked case, I didn't catch it on CI benchmarking data.

Something I like about only benchmarking on release candidates is that a almost statistically insignificant regression will slip by, but add enough together and hopefully the release catches the sum of these regressions in a more noticeable fashion. Similar tactics could be used for other optimizations as well like compile time, binary size, etc.

Reply:
> which in and of itself might be a hard thing to do...do GH action runners suffer from noisy neighbors, etc

Switching to Iai for instruction-count benchmarking would be a way to resolve this

https://bheisler.github.io/criterion.rs/book/iai/iai.html

>  Something I like about only benchmarking on release candidates is that an almost statistically insignificant regression will slip by benchmarks on each commit (because we can't fail a PR for such a small regression that may just be noise), but add enough together and hopefully the release catches the sum of these regressions in a more noticeable fashion. Similar tactics could be used for other optimizations as well like compile time, binary size, etc.

This is a good point that I've learned in the past but forgot over time.  You need to look at **both** item by item but also the bigger picture.  Even better when you have a user-focused target that you are comparing to.